### PR TITLE
[SM-699] Projects bulk delete - add bulk confirmation dialog

### PR DESCRIPTION
**Pull request by Thomas-Avery**
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

The purpose of this PR is to add a new bulk confirmation dialog when the user has read/write and read-only access to the projects selected for bulk delete.

## Code changes

- apps/web/src/locales/en/messages.json

Adding new message for bulk confirmation dialog

- bitwarden_license/bit-web/src/app/secrets-manager/projects/projects/projects.component.ts

Launch the bulk confirmation dialog when the user has a combination of read/write and read-only access.

- bitwarden_license/bit-web/src/app/secrets-manager/shared/dialogs/bulk-confirmation-dialog.component.html
bitwarden_license/bit-web/src/app/secrets-manager/shared/dialogs/bulk-confirmation-dialog.component.ts

New bulk confirmation dialog used to display to the user what items will not be deleted due to permission restrictions. 

- bitwarden_license/bit-web/src/app/secrets-manager/shared/projects-list.component.ts

Change `selectedHasWriteAccess` to return true if any selected project has write access.

- bitwarden_license/bit-web/src/app/secrets-manager/shared/sm-shared.module.ts

Expose the bulk confirmation dialog in the sm-shared module.

### Bulk delete projects when the user has write access to all projects

https://user-images.githubusercontent.com/43214426/231276263-24915e47-6c4a-4d0e-a3ae-6961f11b7f10.mov

### Bulk delete projects when the user has a combination of read-only and write access to projects

https://user-images.githubusercontent.com/43214426/231276358-da08e33c-60a3-41a2-8afc-2e788c216406.mov

### Bulk delete projects when the user has only read-only access to projects

https://user-images.githubusercontent.com/43214426/231276467-92f0cd54-fa66-4ea2-80ff-460802d2eb3f.mov

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

### CODE REVIEW

File: bitwarden_license/bit-web/src/app/secrets-manager/projects/projects/projects.component.ts [hunk 1]
```diff
@@ -70,11 +82,47 @@ export class ProjectsComponent implements OnInit {
...
+          data: {
+            title: "deleteProjects",
+            columnTitle: "projectName",
+            message: "smProjectsDeleteBulkConfirmation",
+            details: this.getBulkConfirmationDetails(readOnlyProjects),
+          },
+        }
+      );
+
+      dialogRef.closed.pipe(takeUntil(this.destroy$)).subscribe((result) => {
```
Comment (willmartian) [inline bitwarden_license/bit-web/src/app/secrets-manager/projects/projects/projects.component.ts:101]: Converting this observable to a promise negates all of the boilerplate of `takeUntil` and `OnDestroy`:

`await lastValueFrom(dialogRef.closed)`

File: bitwarden_license/bit-web/src/app/secrets-manager/shared/dialogs/bulk-confirmation-dialog.component.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  columnTitle: string;
+  message: string;
+  details: BulkConfirmationStatus[];
+}
+
+export class BulkConfirmationStatus {
+  id: string;
+  name: string;
+  description: string;
+}
```
Comment (willmartian) [inline bitwarden_license/bit-web/src/app/secrets-manager/shared/dialogs/bulk-confirmation-dialog.component.ts:15]: Why use a class here? Shouldn't an interface/type suffice?

File: bitwarden_license/bit-web/src/app/secrets-manager/projects/projects/projects.component.ts [hunk 3]
```diff
@@ -70,11 +82,47 @@ export class ProjectsComponent implements OnInit {
...
+            data: {
+              projects: writeProjects,
+            },
+          });
+        }
+      });
+    } else {
+      this.dialogService.open<unknown, ProjectDeleteOperation>(ProjectDeleteDialogComponent, {
+        data: {
+          projects: projects,
```
Comment (willmartian) [inline bitwarden_license/bit-web/src/app/secrets-manager/projects/projects/projects.component.ts:113]: Fun syntax fact, you can just write it as:

```suggestion
          projects,
```

File: bitwarden_license/bit-web/src/app/secrets-manager/projects/projects/projects.component.ts (hunk 1 above)
Comment (willmartian) [reply bitwarden_license/bit-web/src/app/secrets-manager/projects/projects/projects.component.ts:101]: (General rule of thumb: do I only need one value from this subscription? Convert it to a promise.)

**Review by willmartian — CHANGES_REQUESTED** [review]
GG. Couple of small changes requested below:

File: bitwarden_license/bit-web/src/app/secrets-manager/shared/dialogs/bulk-confirmation-dialog.component.ts (hunk 2 above)
Comment (Thomas-Avery) [reply bitwarden_license/bit-web/src/app/secrets-manager/shared/dialogs/bulk-confirmation-dialog.component.ts]: I was using it as a class in the projects.component.ts `getBulkConfirmationDetails`  method.

Swapped to an interface and updated `getBulkConfirmationDetails` https://github.com/bitwarden/clients/pull/5200/commits/1dc59b1aa27b0bf2a810dbda473e9adacf054707

Comment (willmartian) [reply bitwarden_license/bit-web/src/app/secrets-manager/shared/dialogs/bulk-confirmation-dialog.component.ts]: I think we should only use classes when we are expecting to add some functionality to the class. If all we need is build time type checking, lets use types/interfaces. 

Classes stay in the compiled JS output, meaning a larger bundle size and a perf penalty.

**Pull request merged**